musig/src/musig.rs

```diff
@@ -0,0 +1,312 @@
+use crate::hash::{AggregateHash, CommitmentHash, MsgHash, SignatureHash};
+use bellman::pairing::ff::Field;
+use franklin_crypto::eddsa::{PrivateKey, PublicKey, Seed, Signature};
+use franklin_crypto::jubjub::edwards::Point;
+use franklin_crypto::jubjub::{FixedGenerators, JubjubEngine, Unknown};
+
+#[derive(Debug)]
+pub enum MusigError {
+    SelfIndexOutOfBounds,
+    AssigningCommitmentToSelfIsForbidden,
+    DuplicateCommitmentAssignment,
+    AssigningRPubToSelfIsForbidden,
+    AssigningRPubBeforeSettingAllCommitmentsIsForbidden,
+    DuplicateRPubAssignment,
+    RPubDoesntMatchWithCommitment,
+    SigningBeforeSettingAllRPubIsForbidden,
+    SigningShouldHappenOnlyOncePerSession,
+    AggregatingSignatureBeforeSigningIsForbidden,
+}
+
+impl MusigError {
+    pub fn description(&self) -> &str {
+        match *self {
+            MusigError::SelfIndexOutOfBounds => "self index is out of bounds",
+            MusigError::AssigningCommitmentToSelfIsForbidden => {
+                "assigning commitment to self is forbidden"
+            }
+            MusigError::DuplicateCommitmentAssignment => "duplicate commitment assignment",
+            MusigError::AssigningRPubToSelfIsForbidden => "assigning r_pub to self is forbidden",
+            MusigError::AssigningRPubBeforeSettingAllCommitmentsIsForbidden => {
+                "assigning r_pub before setting all commitments is forbidden"
+            }
+            MusigError::DuplicateRPubAssignment => "duplicate r_pub assignment",
+            MusigError::RPubDoesntMatchWithCommitment => "r_pub doesnt match with commitment",
+            MusigError::SigningBeforeSettingAllRPubIsForbidden => {
+                "signing before setting all r_pub is forbidden"
+            }
+            MusigError::SigningShouldHappenOnlyOncePerSession => {
+                "signing should happen only once per session"
+            }
+            MusigError::AggregatingSignatureBeforeSigningIsForbidden => {
+                "aggregating signature before signing is forbidden"
+            }
+        }
+    }
+}
+
+impl std::error::Error for MusigError {}
+
+impl std::fmt::Display for MusigError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        write!(f, "{}", self.description())
+    }
+}
+
+pub struct MusigSession<E: JubjubEngine> {
+    commitment_hash: Box<dyn CommitmentHash<E>>,
+    signature_hash: Box<dyn SignatureHash<E>>,
+    msg_hash: Box<dyn MsgHash>,
+    participants: Vec<PublicKey<E>>,
+    self_index: usize,
+    aggregated_public_key: PublicKey<E>,
+    a_self: E::Fs,
+
+    r_self: PrivateKey<E>,
+    r_pub_aggregated: PublicKey<E>,
+
+    t_others: Vec<Option<Vec<u8>>>,
+    t_self: Vec<u8>,
+    t_count: usize,
+
+    r_pub_others: Vec<Option<PublicKey<E>>>,
+    r_pub_self: PublicKey<E>,
+    r_pub_count: usize,
+
+    performed_sign: bool,
```

Comment:
Instead of having `performed_sign` as a flag, you can simply make `sign` method consuming the `self` by value, thus the whole object won't be usable once the signing operation is performed.

Reply:
That will require to create separate struct for aggregating signature, since after creating a signature MusigSession aggregates signatures from other participants

Reply:
Ah, I see, sorry. My bad.